src/toil/job.py

```diff
@@ -1323,49 +2207,113 @@ def _serialiseFirstJob(self, jobStore):
+        self.description.assignConfig(jobStore.config)
+        
+        return self.description
+        
+    @classmethod
+    def loadJob(cls, jobStore, jobDescription):
+        """
+        Retrieves a :class:`toil.job.Job` instance from a JobStore
 
-    ####################################################
-    #Function which worker calls to ultimately invoke
-    #a jobs Job.run method, and then handle created
-    #children/followOn jobs
-    ####################################################
+        :param toil.jobStores.abstractJobStore.AbstractJobStore jobStore: The job store.
+        :param toil.job.JobDescription jobDescription: the JobDescription of the job to retrieve.
+        :returns: The job referenced by the JobDescription.
+        :rtype: toil.job.Job
+        """
+        
+        # Grab the command that connects the description to the job body
+        command = jobDescription.command
+        
+        commandTokens = command.split()
+        assert "_toil" == commandTokens[0]
+        userModule = ModuleDescriptor.fromCommand(commandTokens[2:])
+        logger.debug('Loading user module %s.', userModule)
+        userModule = cls._loadUserModule(userModule)
+        pickleFile = commandTokens[1]
+
+        # Get a directory to download the job in
+        directory = tempfile.mkdtemp()
+        # Initialize a blank filename so the finally below can't fail due to a
+        # missing variable
+        filename = ''
+
+        try:
+            # Get a filename to download the job to.
+            # Don't use mkstemp because we would need to delete and replace the
+            # file.
+            # Don't use a NamedTemporaryFile context manager because its
+            # context manager exit will crash if we deleted it.
+            filename = os.path.join(directory, 'job')
+                
+            # Download the job
+            if pickleFile == "firstJob":
+                jobStore.readSharedFile(pickleFile, filename)
+            else:
+                jobStore.readFile(pickleFile, filename)
+
+            # Open and unpickle
+            with open(filename, 'rb') as fileHandle:
+                job = cls._unpickle(userModule, fileHandle, requireInstanceOf=Job)
+                # Fill in the current description
+                job._description = jobDescription
+                
+                # Set up the registry again, so children and follow-ons can be added on the worker
+                job._registry = {job.jobStoreID: job}
+                
+                return job
+
+                # TODO: We ought to just unpickle straight from a streaming read
+        finally:
+            # Clean up the file
+            if os.path.exists(filename):
+                os.unlink(filename)
+            # Clean up the directory we put it in
+            # TODO: we assume nobody else put anything in the directory
+            if os.path.exists(directory):
+                os.rmdir(directory)
```

Comment:
```
            # TODO: we assume nobody else put anything in the directory
            if os.path.exists(directory):
                os.rmdir(directory)
```

I think it might be worth preemptively using `shutil.rmtree` here as it might (or might not) prevent this code's spirit from reincarnating later as a bug and I see no downside to doing so.

Reply:
OK, I'll do that.